### DISCUSSION TITLE
v0.23.0-rc2: app.OnServe().Add undefined (type *hook.Hook[*core.ServeEvent] has no field or method Add)compilerMissingFieldOrMethod

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#5610] [Q&A] [answered]
I'm trying to migrate a static file route

<details><summary>v0.22.21</summary>

```go
func main() {
	app := pocketbase.New()

	migratecmd.MustRegister(app, app.RootCmd, migratecmd.Config{
		Automigrate: true,
	})

	// Add base path to serve ui app
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {

		e.Router.GET(
			"/*",
			apis.StaticDirectoryHandler(ui.DistDirFS, false),
		)
		return nil
	})

     if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

</details> 

Following the instructions [here](https://pocketbase.io/v023upgrade/go/#static-files-serving), so now: 
```go
// Add base path to serve ui app
	app.OnServe().Add(func(se *core.ServeEvent) error {

		se.Router.GET(
			"/*",
			apis.Static(ui.DistDirFS, false),
		)
		return se.Next()
	})
```

and getting the following error: 

`app.OnServe().Add undefined (type *hook.Hook[*core.ServeEvent] has no field or method Add)compiler[MissingFieldOrMethod]`

**Comment by ganigeorgiev** [comment, marked as answer]
The `Add` hooks method is replaced by [`Bind`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc2/tools/hook#Hook.Bind) and [`BindFunc`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc2/tools/hook#Hook.BindFunc).

The above code in v0.23.0 should look like:

```go
app.OnServe().BindFunc(func(se *core.ServeEvent) error {
    // note that Static requires a named "path" wildcard
    // (see https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc2/apis#Static)
    se.Router.GET("/{path...}", apis.Static(ui.DistDirFS, false))

    return se.Next()
})
```

_(https://pocketbase.io/v023upgrade/go/#static-files-serving)_

**Reply by ganigeorgiev** [reply]
_Oh sorry, on second read, it seems that I forgot to rename the method in the upgrade guide. ~Will update the documentation a little bit later~ Done._

**Reply by Plorenzo** [reply]
that makes more sense! thanks!

**Comment by dschissler** [comment]
@ganigeorgiev I'm going to hijack thread this to mention an upgrade docs typo.

I think that there might be two mistakes here.  Perhaps `:name` should be `{name}` and `c.JSON` should be `e.JSON`.

```golang
app.OnServe().BindFunc(func(se *core.ServeEvent) error {
    se.Router.GET("/hello/:name", func(e *core.RequestEvent) error {
        name := e.Request.PathValue("name")

        return c.JSON(http.StatusOK, map[string]string{
            "message": "Hello " + name,
        })
    }).Bind(apis.RequireAuth())

    return se.Next()
})
```

Possibly (I'm not 100% sure) there are several mistakes where the right hand (new side) says `c.Request` when it should be `e.Request`.

Also in jvm upgrade docs `c.request` -> `e.request`.

**Reply by dschissler** [reply]
@ganigeorgiev 

Should be lowercase `static`:

`routerAdd("GET", "/{path...}", $apis.Static("/path/to/public", false))`
`routerAdd("GET", "/{path...}", $apis.static("/path/to/public", false))`

**Reply by denzuko** [reply]
@dschissler lets note that most of the answers here are talking about golang and your reply seems to be in, javascript?

Golang members use case for public/private scoping since golang does not use polymorphisms where javascript does and uses function scoping for member access.

**Reply by dschissler** [reply]
@denzuko Let's note that Pocketbase provides an API for both Golang and Javascript (as a framework).  Where did you come from on the Internet to be so helpful?

**Reply by denzuko** [reply]
> @denzuko Let's note that Pocketbase provides an API for both Golang and Javascript (as a framework). Where did you come from on the Internet to be so helpful?

not sure if that's an earnest question or sarcasm but to answer it, a thing called Fidonet and 2600's irc. Plus being a code contributor to several major golang based projects including the Go Foundation.

The point being, that yes Pocketbase does provides an API for both **Golang** and Javascript. But for the google dorks / LLM bots that come along later its best to understand there's a different approach need to take because of the differences in the two languages even though those api calls may be similar. But I'm sure yourself already knew that.
